### PR TITLE
rpcserver: add optional constraint fields to AddInvoice

### LINKED ISSUE #2004: [feature]: RFQ: First-Class Limit Orders In Quote Negotiation
## Summary
Today, RFQ orders are mostly constrained by:
- Requested max size (`asset_max_amt` or `payment_max_amt`)
- Quote expiry
- Channel availability and balance shape
- Price-oracle tolerance checks

This is sufficient for basic RFQ flows, but it is not a full limit-order model. We should add explicit limit-order semantics (price bounds, fill constraints, and execution policy) to quote negotiation.

## Why This Matters
Limit-order semantics enable:
- Price protection: avoid execution outside user-defined acceptable price bounds.
- Slippage control: reduce risk from stale hints, spread moves, and short-term volatility.
- Deterministic UX: users can express explicit execution intent instead of best-effort behavior.
- Safer automation: wallet agents can auto-negotiate with tighter risk controls.
- Better policy explainability: clearer machine-readable reject reasons.
- Better multi-peer execution: easier to split across peers while preserving user constraints.

## Terminology
- `IOC` (`Immediate-Or-Cancel`): execute immediately for any amount that satisfies constraints; cancel any unfilled remainder.
- `FOK` (`Fill-Or-Kill`): execute immediately and fully, or cancel the entire order.

### What RFQ can express today
- Buy flow: max asset amount (`AssetMaxAmt`) + expiry + peer + metadata.
- Sell flow: max payment amount (`PaymentMaxAmt`) + expiry + peer + metadata.
- Request/accept negotiation primarily settles on a rate + expiry.

### What is enforced in practice
- Requester-side quote validation checks include quote expiry bounds and oracle tolerance against the accepted rate.
- HTLC/policy enforcement is max-cap based, where sale path policy derives from `AssetMaxAmt` and purchase path policy derives from `PaymentMaxAmt`.
- Liquidity constraints are effectively channel-driven: channel existence checks gate order submission, and channel balances influence selection/sorting.

### Existing but underused protocol surface
- Wire format already includes optional `MinInAsset` and `MinOutAsset` fields.
- These min fields are not first-class in typed request structs and RPC order APIs.

### Current gap
Current behavior is closer to:
- "Market quote with max-size cap + expiry"
not:
- "Limit order with explicit price/fill constraints"

Users cannot directly encode:
- Minimum acceptable price / rate.
- Maximum acceptable price / rate.
- Minimum fill size.
- Execution policy (`IOC` vs `FOK`).

### Workstream A: Add explicit limit constraints to request surfaces
- Add optional limit-price/rate bound fields to buy/sell order APIs.
- Add optional fill bound fields (`min_in_asset`, `min_out_asset`) to typed requests and RPC.
- Keep current max fields as-is for compatibility.
- Define a canonical rate-bound representation that avoids buy/sell direction ambiguity.

### Workstream B: Enforce limits during negotiation
- Responder side (`ResolveRequest`): reject when request limits cannot be satisfied.
- Requester side (`VerifyAcceptQuote`): enforce explicit requester limits in addition to oracle tolerance.
- Return explicit reject reasons for limit failures (price bound miss, min fill miss, insufficient liquidity).

### Workstream C: Add execution policy semantics
- Support explicit execution policy on orders.
- `IOC`: allow partial immediate execution above minimum fill.
- `FOK`: require full immediate execution or reject.
- Ensure policy is respected consistently in quote acceptance and HTLC policy installation.

### Workstream D: Negotiate explicit fill quantities
- Add accepted fill quantity semantics to quote acceptance path.
- Bind policy persistence and HTLC enforcement to negotiated fill values, not only request max caps.
- This removes ambiguity around partial fills and improves deterministic execution.

### Workstream E: Multi-peer selection that respects limits
- When multiple peers are available, rank and split using effective rate and usable capacity.
- Ensure every shard respects the same order limits and execution policy.

## Non-Goals
- Building a centralized order book.
- Long-lived resting/GTC orders in initial implementation.
- Redesigning the external price-oracle architecture.

## Compatibility Expectations
- New fields should be additive and optional.
- Old peers should continue to interoperate under current behavior.
- Limit-specific validation should only activate when limit fields are present.
- Use versioning only where strict incompatibility requires it.

## Acceptance Criteria
- Users can specify explicit limit-price bounds for buy and sell RFQ orders.
- Users can specify minimum fill constraints.
- Users can choose execution policy (`IOC` or `FOK`).
- Quotes violating constraints are rejected with clear machine-readable reasons.
- Accepted quotes cannot pass validation if they violate requester limits.
- Policy/HTLC enforcement is tied to negotiated constraints, not only broad request max caps.
- Test coverage includes buy and sell flows.
- Test coverage includes `IOC` and `FOK` behaviors.
- Test coverage includes backward compatibility with old peers.
- Test coverage includes volatility and timing edge cases.

## Open Questions
- What is the canonical limit-price representation to avoid buy/sell interpretation errors?
- Should both min and max rate bounds be supported initially, or only one directional bound per order type?
- Do we extend `QuoteRespStatus`, add new reject codes, or both?
- Is explicit fill quantity required in the first delivery, or can it follow once request-side limits are live?
- Should passive offer filtering and limit-order checks be merged into one decision pipeline?

### PR DESCRIPTION
**Pull request by jtobin**
Depends on #2050.

Adds the limit order constraint fields (`asset_min_amt`, `asset_rate_limit`, and `execution_policy`) to the AddInvoice RPC, such that invoice quotes can be negotiated using limit order semantics. The fields are all optional; without them we just proceed with "vanilla" RFQ negotiation.

### CODE REVIEW

File: rfq/portfolio_pilot_test.go [hunk 1]
```diff
@@ -450,6 +461,476 @@ func TestResolveRequest(t *testing.T) {
...
+					func(e rfqmsg.RejectErr) {
+						require.Equal(
+							t,
+							rfqmsg.PriceBoundMissRejectCode, //nolint:lll
+							e.Code,
+						)
+					},
+				)
+			},
+		},
```
[the remark this answers is not in the document]
Comment (jtobin) [reply rfq/portfolio_pilot_test.go:671]: Resolved.

File: itest/rfq_test.go [hunk 2]
```diff
@@ -842,6 +1118,1098 @@ func testRfqPortfolioPilotRpc(t *harnessTest) {
...
+		require.NoError(t.t, err)
+
+		_, ok := event.Event.(*rfqrpc.RfqEvent_PeerAcceptedBuyQuote)
+		require.True(t.t, ok, "expected PeerAcceptedBuyQuote, "+
+			"got: %v", event)
+	}, rfqTimeout)
+
+	err = carolEvents3.CloseSend()
+	require.NoError(t.t, err)
 }
```
[the remark this answers is not in the document]
Comment (jtobin) [reply itest/rfq_test.go]: Resolved.

**Comment by litbot-9000** [comment]
@jtobin, remember to re-request review from reviewers when ready

File: docs/release-notes/release-notes-0.8.0.md [hunk 3]
```diff
@@ -403,6 +410,19 @@
 - [PR#2049](https://github.com/lightninglabs/taproot-assets/pull/2049):
   Add unit, property-based, and integration tests for execution policy.
 
+- [PR#2050](https://github.com/lightninglabs/taproot-assets/pull/2050):
+  Add unit and integration tests for fill quantity negotiation,
+  including wire roundtrip, zero-normalisation, fill-vs-constraint
+  validation, sell-side FOK/IOC with fill caps, and responder
+  constraint rejection.
+
+- [PR#2053](https://github.com/lightninglabs/taproot-assets/pull/2053):
```
Comment (darioAnongba) [inline docs/release-notes/release-notes-0.8.0.md:405]: usual changelog comment about unnecessary entries.

File: docs/release-notes/release-notes-0.8.0.md [hunk 4]
```diff
@@ -403,6 +410,19 @@
 - [PR#2049](https://github.com/lightninglabs/taproot-assets/pull/2049):
   Add unit, property-based, and integration tests for execution policy.
 
+- [PR#2050](https://github.com/lightninglabs/taproot-assets/pull/2050):
```
Comment (darioAnongba) [inline docs/release-notes/release-notes-0.8.0.md:399]: rebase issue? this references 2050. But unnecessary in any case.

File: rpcserver/invoiceamount_test.go [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+		ValueMsat: 7_000_000,
+	}
+
+	_, err := validateInvoiceAmount(quote, 0, inv)
+	require.ErrorContains(t, err, "max routable amount")
+}
+
+// TestCapEffectiveAssetAmount verifies the capping logic that
+// guards validateInvoiceAmount in AddInvoice.
+func TestCapEffectiveAssetAmount(t *testing.T) {
```
Comment (darioAnongba) [inline rpcserver/invoiceamount_test.go]: tautology test? this is testing nothing. You can extract `capFillAmount` from `rpcserver` and test it independently (also saves code duplication if reused). Can also just delete this test I think since cap is already tested in `TestValidateInvoiceAmountPartialFill`

File: taprpc/tapchannelrpc/tapchannel.proto [hunk 6]
```diff
@@ -251,6 +251,21 @@ message AddInvoiceRequest {
...
+    // Forwarded to AddAssetBuyOrder. 0 means unset.
+    uint64 asset_min_amt = 9;
+
+    // Optional rate floor (asset units per BTC) as a
+    // fixed-point number. Forwarded to AddAssetBuyOrder.
+    rfqrpc.FixedPoint asset_rate_limit = 10;
+
+    // Execution policy (IOC default, FOK). Forwarded to
+    // AddAssetBuyOrder.
+    rfqrpc.ExecutionPolicy execution_policy = 11;
```
Comment (darioAnongba) [inline taprpc/tapchannelrpc/tapchannel.proto:270]: Thinking about this a bit, I believe there is a gap? Currently the default enforces nothing when I believe users expect the default to be: "if the exchange succeeds, I get `AssetAmt`". By default I mean when the user specifies nothing (no execution policy = IOC, `asset_min_amt = 0`). Logically I see either:
- Default Execution Policy = FOK so it's the defined `AssetAmt` or nothing
- (better) Default Execution Policy = IOC but `AssetMinAmt = AssetAmt`.

To achieve this, we need to differentiate between "not set" and zero, so the field becomes: `optional uint64 asset_min_amt = 9;`

then, if `asset_min_amt = nil`, we clamp to `AssetAmt`. For `asset_min_amt` to be 0, the user needs to set it explicitely.

It's the good old Go/proto problem of undefined = zero/empty, instead of undefined = nil.

File: rpcserver/invoiceamount_test.go (hunk 5 above)
Comment (jtobin) [reply rpcserver/invoiceamount_test.go]: Nice catch, will axe.

File: taprpc/tapchannelrpc/tapchannel.proto (hunk 6 above)
Comment (jtobin) [reply taprpc/tapchannelrpc/tapchannel.proto:270]: _Very_ nice catch. Completely agree, will fix.

File: docs/release-notes/release-notes-0.8.0.md (hunk 3 above)
Comment (jtobin) [reply docs/release-notes/release-notes-0.8.0.md:405]: I'll maintain these for now but chop them out entirely prior to v0.8.0. In the future we should get rid of many of these template sections entirely.

File: docs/release-notes/release-notes-0.8.0.md (hunk 4 above)
Comment (jtobin) [reply docs/release-notes/release-notes-0.8.0.md:399]: Axed in latest changeset.

File: rpcserver/invoiceamount_test.go (hunk 5 above)
Comment (jtobin) [reply rpcserver/invoiceamount_test.go]: Axed in latest changeset.

File: taprpc/tapchannelrpc/tapchannel.proto (hunk 6 above)
Comment (jtobin) [reply taprpc/tapchannelrpc/tapchannel.proto:270]: Resolved (according to the straightforward differentiation scheme you describe) in latest changeset.

**Review by darioAnongba — APPROVED** [review]
LGTM
